### DISCUSSION TITLE
Script Kit 2.0 Release Candidate

### DISCUSSION THREAD

**Discussion by johnlindquist** [johnlindquist/kit#1365] [Announcements]
# Script Kit 2.0 Release Candidate

Download for your platform from the releases page:
https://github.com/johnlindquist/kitapp/releases/tag/v1.99.82

## New Main Menu

The main menu has been re-designed and optimized for your keyboard-centric workflows.

<img width="827" alt="CleanShot 2023-10-27 at 07 47 09@2x" src="https://github.com/johnlindquist/kit/assets/36073/60b9d227-d0a1-4cf0-ae2a-bede13fba9ef">

## New Custom OSX Window

Prompts like `editor` and `term` which use `ignoreBlur` will now convert the window into an "app" window which can be accessed by `cmd+tab`, is pinned when switching spaces, and shows up in Mission Control. This is also forward-looking to v3 which I'll talk about more in the future...

## Keywords
Borrowed directly from Alfred, keywords allow you to trigger a script by typing a keyword followed by a space.

A keyword can be a single character or a word. For example, `c` for `clipboard`.

### Built-in Keywords
- `s` - Displays Snippets from your Snippets Directory
- `c` - Displays Clipboard History
- `f` - Find script by searching script contents
- `kit` - Access kit settings
- `kenv` - Access kenv options
- `npm` - Add/remove npm packages
- `spell` - List spelling suggestions for the input
 
 And many others. To see them all, type `keyword` into the Script Kit prompt.

 To create a custom keyword for your own scripts, add:

 ```
 // Keyword: mycustomkeyword
 ```

 If the first prompt is an `await arg()`, it will display the list in the main menu. Otherwise, it will jump to the next prompt.

 Also, if the first prompt is a static list and you shave off the few milliseconds required to load the list from your script, you can cache the list (after the first run) by adding:

 ```
 // Cache: true
 ```

Use caching sparingly, as it's only useful for `// Keyword` and `// Shortcut` where the first prompt is an `arg` with a static list.

## Pass Main Menu Input to a Script

```
// Pass: true
```

The `Pass` metadata allows you to type into the main menu without your script being filtered out of the list. Once you select the script, the input you've typed will be "passed" into the script. You can access the "passed" input using `flag?.pass`. For example, if you want to pass the main menu input to the input of you first arg:

```
// Name: Pass Demo
// Pass: true

import "@johnlindquist/kit"

await arg({
  input: flag?.pass || "",
})
```

### Pass "postfix"

Image you want to type: "How many moons does Jupiter have?" and have the "?" automatically trigger a script that reaches out to an AI service. You can do this by adding a "postfix" to the `Pass` metadata.

```
// Pass: ?
```

This will automatically trigger the script when you type a "?" at the end in the main menu and the entire input will be passed to the script as `flag?.pass`.

## Multi-select Prompt `await select()`

<img width="833" alt="CleanShot 2023-10-27 at 15 26 56@2x" src="https://github.com/johnlindquist/kit/assets/36073/9420d7a8-096c-468b-b98c-a9f70b48569b">

```js
let arrayOfChoices = await select("Pick one or more", ["one", "two", "three"])
```

Probably the most requested feature, Script Kit now has a `select` prompt which allows you to pick multiple choices.

## Themes Per Light/Dark Mode

You can now customize which theme will be active for both light and dark more or use the same theme for both.

## Snippets Directory

In your ~/.kenv/snippets directory, create .txt files you want to use as snippets. For example:
1. Create `~/.kenv/snippets/lorem.txt`
2. Add the following text to the file:
```
// Name: Lorem Ipsum
Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua.
```

The "Lorem Ipsum" snippet will now appear in the Snippets menu using the `s` keyword. Select the snippet will paste it into the current application.

Adding Snippet metadata will allow you to invoke the snippet anywhere on your system. It's a best practice to use a "postfix", such as ",,", to avoid triggering the snippet when you don't want to.
```
// Snippet: lorem,,
```

### Snippet Creator

<img width="829" alt="CleanShot 2023-10-27 at 01 02 25@2x" src="https://github.com/johnlindquist/kit/assets/36073/a32d5e12-8ff2-441c-8edf-cfdfe3b1ab0d">

Typing `ns ` from the main menu will launch the quick snippet creator. There's a helpful guide there that will show all the available options. Just to list a few:

- Tab stops: $0, $1, $2, $3
- Dropdown list: ${|apple,banana,cherry|}
- Current selection: $SELECTION
- Clipboard content: $CLIPBOARD

and many more...

If you need more advanced scenarios, for example clipboard history:

```js
let history = await getClipboardHistory()
let text = await arg("Pick an item from your history", history.map(item => item.value)

let result = await template(`Hello $0,

${text}
`)

await writeFile(home("example.txt"), result)
```

## Actions

The new `actions` array makes it much easier to assign shortcuts and behaviors when you have a choice selected. You can access the actions menu by pressing `right`on the keyboard, then quickly search to execute the action you want to take with the currently focused choice:

```js
// Name: Actions Example

import { Action } from "@johnlindquist/kit"

// Using a "flag" determines where to do to custom logic: After the prompt or in the action
// Also, "flags" are supported when running the script in the terminal with `--js`
let actions: Action[] = [
  {
    shortcut: `${cmd}+t`,
    name: "Append .ts",
    visible: true, // Display shortcut in the prompt
    onAction: async (input, state) => {
      // Since we're not using a "flag", we can do custom logic here
      submit(`${state.focused?.name}.ts`)
    },
  },
  {
    name: "Append .js",
    flag: "js", // Set `global.flag.js` to true when selecting the action
  },
]

let result = await arg("Pick a number", ["one", "two", "three"], actions)

if (flag.js) {
  result = `${result}.js`
}

inspect(result)
```

## Previews on All Prompts

You can build-out helpful previews on the right side for all prompts now (see the Snippet Creator above for a great example). This is especially useful for prompts like `editor`, `drop`, and `term` where you may need reminders on what the script is doing or steps you want to guide the user through.

```js
await editor({
  value: `Book Outline`,
  preview: md(`# Requirements
- Cover Major Characters
- Address the theme
- Have a beginning, middle, and end  
  `),
})
```

## Move kenv to a Different Directory

Type `kit move` into the main prompt to move your kenv directory to a different location. This is useful if you want to keep your scripts in dotfiles or other setup.

## onClick and onKeydown Globals
- Global event handlers for click and type events.

## Custom TypeScript Loader

The new custom loader eliminates the need to watch .ts files for changes (which caused many issues from a long-running app). We now have a customized loader that can run any TypeScript file instantly.

An awesome benefit of this is you can create a script anywhere on your system, such as `~/Desktop/my-script.ts` (or js if you prefer) and run it from script runner: `~/.kit/bin/kit ~/Desktop/my-script.ts`.

> Note: Your IDE won't recognize the `import` statements since it's not in a project, but the script will run just fine.

If we get enough requests, we will allow you to run the scripts from the built-in file browser through the app as well.

## Watch Scripts for npm Install

As you're writing your scripts in your favorite editor, Kit.app will detect if you've added a new npm package and automatically prompt you to install it using the Kit.app built-in terminal.

## Road to v2

The plan is to release v2 in mid-December. I'll be focusing on updating documentation, guides, tips, and tricks to finally help everyone get a more complete grasp of what Script Kit is capable of. There are a few more APIs which haven't been announced which will be included as well.

## Questions?

Please ask below. We're always happy to help figure out better workflows to optimize your day. Happy scripting!

\- John

**Comment by josephcrawfordSRH** [comment]
@johnlindquist I upgraded to this and it would not allow my script to work any longer, the async function that returned an array to the arg method would no longer allow me to search.  I quickly went to the site and downloaded the older version I had and once installed everything seems to have broken.

![Screenshot 2023-11-14 at 8 48 39 AM](https://github.com/johnlindquist/kit/assets/100154037/ab52f74c-ae53-4558-abbb-0bf4691f94a6)

I thought that using AppZapper to remove all associated files and removing my .kenv/ directory would have solved this but even after a reinstall that is not happening.  Upon first open I get a t.map is not a function call. 

[
<img width="773" alt="Screenshot 2023-11-14 at 8 55 29 AM" src="https://github.com/johnlindquist/kit/assets/100154037/a7bef0d1-e3dd-4934-8967-cdd038e13782">
](url)

**Reply by josephcrawfordSRH** [reply]
I found that I had to remove .kenv, .kit and .knode to reinstall and get it working again.  I will have to look into why my script was not working with 2.0 RC when I have more time.

**Reply by johnlindquist** [reply]
Can you share the script with me publicly/privately? I'd love to debug it

> Note: 99% of the time, the `~/.kit` SDK is the cause of updating issues where a file/folder was open or something and prevented a successful upgrade. The only issues people ever have with `~/.kenv` is the node_modules dir, there should never be any need to remove it. It's just script files.

**Reply by josephcrawfordSRH** [reply]
@johnlindquist I have stripped out everything work related at this point so I do not mind sharing publicly as others may find this useful.  I do plan to refactor this quite a bit but wanted to get it working so that i could use it for faster productivity in the meantime.

In order to test this properly you should have a few git projects under a single directory.  The first question will be about vscode but the second should ask for a path to your projects directory.

```typescript
// opob.js
// Exclude: true

import "@johnlindquist/kit";
type ICache = {
  projectDir: string;
  editor: string;
  projects: string[];
  branches: {
    [key: number]: string[];
  };
}

const cyrb53 = (str: string, seed = 0) => {
  let h1 = 0xdeadbeef ^ seed, h2 = 0x41c6ce57 ^ seed;
  for (let i = 0, ch; i < str.length; i++) {
    ch = str.charCodeAt(i);
    h1 = Math.imul(h1 ^ ch, 2654435761);
    h2 = Math.imul(h2 ^ ch, 1597334677);
  }
  h1 = Math.imul(h1 ^ (h1 >>> 16), 2246822507);
  h1 ^= Math.imul(h2 ^ (h2 >>> 13), 3266489909);
  h2 = Math.imul(h2 ^ (h2 >>> 16), 2246822507);
  h2 ^= Math.imul(h1 ^ (h1 >>> 13), 3266489909);

  return 4294967296 * (2097151 & h2) + (h1 >>> 0);
};

const OPOB_CACHE_FILE = kenvPath('cache/opob.json');
let cache = await getCache();
let code = await getCodeEditor();

export async function opob(newWindow = false) {
  if (!cache.projectDir) {
    cache.projectDir = await path(home());
  }

  let project = await arg("Which project?", async () => {
    let projects = await getCacheKey('projects');
    if (!projects) {
      const p = await exec(
        `find -L ${cache.projectDir} -maxdepth 4 -name .git -type d -prune`,
      );
      const pList = p.stdout.split("\n");

      projects = pList.map((project) => {
        return project.replace(cache.projectDir + "/", "").replace("/.git", "");
      });

      await cacheProjects(projects);
    }

    return projects;
  });

  const projectPath = `${cache.projectDir}/${project}`;

  let stashed = false;

  await cd(projectPath);
  const changes = await exec(`git diff --name-only`).stdout;

  const branch = await arg("Which branch?", async () => {
    let branches = cache.branches || {};
    let projectHash = cyrb53(project);

    if (!branches[projectHash]) {
      const b = await fetchBranchesFromServer(projectPath);

      branches[projectHash] = b;
      await cacheBranches(branches);
    }

    return currentBranchFirst(branches[projectHash]);
  });


  let windowFlag = "-r";
  if (newWindow) {
    windowFlag = "-n";
  }

  await cd(projectPath);
  if (changes && branch !== await getCurrentBranch()) {
    if (
      await arg("You have changed files, do you want to stash them?", [
        "Yes",
        "No",
      ]) === "Yes"
    ) {
      stashed = true;
      await exec(`cd ${projectPath} && git stash push -m "OPOB Stash"`);
    }
  }

  if (!changes || (changes && stashed)) {
    await exec(`git checkout ${branch}`);
  }

  await exec(`${code} ${windowFlag} ${projectPath}`);
}

async function getCodeEditor(): Promise<string> {
  let code = cache.editor;
  if (!code) {
    let vscodes = [];

    const vscodeinsiders = await $`command -v code-insiders`;
    if (vscodeinsiders.stdout) {
      vscodes.push("code-insiders");
    }

    const vscode = await $`command -v code`;
    if (vscode.stdout) {
      vscodes.push("code");
    }

    if (vscodes.length > 1) {
      code = await arg("Which VSCode?", vscodes);
      if (await arg("Remember VSCode?", ["Yes", "No"]) === 'Yes') {
        // cache the code they wish to use.
        cache.editor = code;
      }
    }

    if (vscodes.length === 0) {
      let message = md(`# VSCdoe CLI not found
    
\`Visual Studio Code\` CLI was not found on your PATH accessible in this script.

Please check \`code\` is installed.

Options:
- Install the VSCode CLI
  - In VSCode hit \`cmd+shift+p\` and type \`shell\` and select \`Shell Command: Install 'code' command in PATH\`.
    `)
      await div(message);
      exit(1);
    }
  }

  return code;
}

async function getCache(): Promise<ICache> {
  return JSON.parse(await ensureReadFile(OPOB_CACHE_FILE, "{}"));
}

async function cacheProjects(projects) {
  cache.projects = projects;
  await writeFile(OPOB_CACHE_FILE, JSON.stringify(cache));
}

async function cacheBranches(branches) {
  cache.branches = branches;
  await writeFile(OPOB_CACHE_FILE, JSON.stringify(cache));
}

async function fetchBranchesFromServer(projectPath) {
  const branchString = await exec(
    `cd ${projectPath} && git fetch --all && git branch -a`,
  );

  let branches = branchString.stdout.split("\n");
  branches = branches.filter((b) => {
    return b !== "" && !b.includes("dependabot") &&
      !b.includes("HEAD ->") && !b.includes("* ") &&
      !branches.includes(b.replace("remotes/origin/", "").trim())
  }).map((b) => {
    return b.replace("remotes/origin/", "").trim();
  });

  return [...branches];
}

async function getCacheKey(key) {
  const cache = ensureReadFile(OPOB_CACHE_FILE, "{}");
  return cache[key];
}

async function getCurrentBranch() {
  const current = await exec(`git symbolic-ref --short HEAD`);
  return current.stdout.trim();
}

async function currentBranchFirst(branches) {
  const currentBranch = await getCurrentBranch();

  const filteredBranches = branches.filter(b => b !== currentBranch);

  return [currentBranch, ...filteredBranches];
}
```

```typescript
// => opobcw.js
// Name: Open on branch current window
// Shortcut: cmd+;

import "@johnlindquist/kit";

import { opob } from "./opob.ts";

await opob(false);
```

```typescript
// => opobnw.js
// Name: Open on branch in new window
// Shortcut: shift+cmd+;

import "@johnlindquist/kit";

import { opob } from "./opob.ts";

await opob(true);
```

**Reply by josephcrawfordSRH** [reply]
One thing you may find is that I remapped my kit shortcut to be `ctrl+space` as the `cmd+;` I am using for my script shortcut.

**Reply by johnlindquist** [reply]
@josephcrawfordSRH So what's the current error?

1. Main shortcut, broken?
2. Main shortcut, run script, broken?

**Reply by johnlindquist** [reply]
Also, in a lib, it's best to avoid top-level await. Those will be immediately executed on import.

I also recommend avoiding `zx` aka `$` (I need to change that in the guide, it was just a highly requested feature a while back). 

`exec` is better at reporting a proper error when those commands fail.

**Reply by josephcrawfordSRH** [reply]
When I tried this under 2.0 RC the projects would load but searching them would show an empty list, removing the search it remained an empty list until I opened the command again.

On 1.79.5 all is working fine.

**Reply by josephcrawfordSRH** [reply]
> Also, in a lib, it's best to avoid top-level await. Those will be immediately executed on import.

In my case I want them to be executed when the function is called because when I hit the keyboard command the function is called, it will then prompt me for certain things based on the cache having certain values or not.

**Reply by johnlindquist** [reply]
@josephcrawfordSRH These are the top-level awaits I was referring to:

```
let cache = await getCache();
let code = await getCodeEditor();
```

I'll run your scripts and see if I can reproduce

**Reply by josephcrawfordSRH** [reply]
@johnlindquist so if I understand you correctly you are saying those two awaits should be done in the actual scripts and the values passed into the base?

**Reply by johnlindquist** [reply]
awaits should execute inside the context they're needed. They're currently at the module-level and are immediately invoked just by importing them.

Since this script is being treated as a library (I use the term "library" to mean "a collection of exported function"), I would move them inside the exported functions so they're only called if that function is called.

I would use your suggestion if I wanted a different cache per script.

Mostly, it's to be aware of "when" your code is executing because if it fails at the time of import, it won't even make it to any of your code/prompts.

This isn't a hard rule, just a suggestion based on my experience.

**Reply by josephcrawfordSRH** [reply]
I see what you mean, the issue is that functions below the exported function which are called internally need access to the cache.  To get the best of both worlds I may just define the variables outside of the function and have the calls done inside of the function.

**Reply by johnlindquist** [reply]
@josephcrawfordSRH any chance you can post a screen recording so I'm absolutely clear of the behavior?

**Reply by johnlindquist** [reply]
@josephcrawfordSRH I believe I see what's going on:

Using a function with arg implies that you want to implement your own custom searching/generation logic. Typically, that would look like this:

```js
let choices = ["apple", "banana", "orange"]
await arg("Type something", async (input)=> {
return choices.filter(c => c.includes(input)) //custom searching/generation logic based on input
})
```

This function will run _every time_ a key is pressed. So the logic in your script is running `find`/`fetch` with every key press, even though it's not changing based on input.

### bug
2.0 refactored the searching/scoring logic. There's now a check if there's no input argument, it only run the first pass, then switches to a noop. This is causing the list to clear when you type. I'll re-implement a "run every time, even if no input", but...

### solution
 I strongly recommend refactoring by hoisting the async functions outside of your `arg`s then call them before the `arg` runs:

```js
let getAllBranches = async ()=> { // branches logic } // just copy/paste the async function
let branches = await getAllBranches()
let branch =await arg("Which branch?", branches)

// Also do the same with the "projects" logic
```

Please let me know if something's not clear.


#### Also

I recommend using `cwd` inside the `exec` options, instead of `cd`:

```bash
let result = await exec(`ls`, {
  cwd: home("dev"),
})

inspect(result.stdout)
```

It helps avoid accidentally running commands in the wrong directly if you ever refactor of if `cd` is called somewhere else in your script.

**Reply by josephcrawfordSRH** [reply]
Awesome thanks for the feedback I will see what I can work out later this evening or tomorrow evening. I appreciate your ability to respond so quickly.

**Reply by josephcrawfordSRH** [reply]
@johnlindquist I refactored the code quite a bit last night and upgraded to v2 as suggested.  Things are working somewhat smooth but when refactoring to use your cwd suggestion things are not quite working.

```javascript
async function fetchBranchesFromServer(project) {
  let projectDir = await getProjectDirectory();
  console.log(`${projectDir}/${project}`); // this is a path /Users/username/projects/project which is a git repo

  // These were originally combined with `git fetch --all && git branch -a` but it was failing so i figured i would try to separate them.
  await exec(`git fetch`, {
   cwd: `${projectDir}/${project}`, 
  });

  const branchString = await exec(`git branch -a`, {
    cwd: `${projectDir}/${project}`,
  });

  let branches = branchString.stdout.split("\n");
  branches = branches.filter((b) => {
    return b !== "" && !b.includes("dependabot") &&
      !b.includes("HEAD ->") && !b.includes("* ") &&
      !branches.includes(b.replace("remotes/origin/", "").trim())
  }).map((b) => {
    return b.replace("remotes/origin/", "").trim();
  });

  return [...branches];
}
```

```text
Command failed with exit code 128: git symbolic-ref --short HEAD
fatal: not a git repository (or any of the parent directories): .git
Error: Command failed with exit code 128: git symbolic-ref --short HEAD
fatal: not a git repository (or any of the parent directories): .git
    at makeError (file:///Users/josephcrawford/.kit/node_modules/@johnlindquist/globals/dist/index.js:36:111645)
    at file:///Users/josephcrawford/.kit/node_modules/@johnlindquist/globals/dist/index.js:36:127769
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
```

I have verified that the `${projectDir}/${project}` does in-fact point to the project directory that is a git repo.

When the git commands were combined, this is the error that I was getting

```text
Command failed with exit code 128: git fetch && git branch -a
fatal: '&&' does not appear to be a git repository
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
Error: Command failed with exit code 128: git fetch && git branch -a
fatal: '&&' does not appear to be a git repository
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
    at makeError (file:///Users/josephcrawford/.kit/node_modules/@johnlindquist/globals/dist/index.js:36:111645)
    at file:///Users/josephcrawford/.kit/node_modules/@johnlindquist/globals/dist/index.js:36:127769
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
```

**Reply by josephcrawfordSRH** [reply]
@johnlindquist This is the full source 

```javascript
// Exclude: true

import "@johnlindquist/kit";

export async function opob(newWindow = false) {
  let editor = await getCodeEditor();

  const projectDir = await getProjectDirectory();
  const projects = await getProjects();

  const project = await arg("Which project?", projects);
  const projectPath = `${projectDir}/${project}`;

  const branches = await getAllProjectBranches(project);
  const branch = await arg("Which branch?", branches);

  let windowFlag = "-r";
  if (newWindow) {
    windowFlag = "-n";
  }

  let changes = await exec(`git diff --name-only`, {
    cwd: projectPath,
  });

  if (changes && branch !== await getCurrentBranch()) {
    if (
      await arg("You have changed files, do you want to stash them?", [
        "Yes",
        "No",
      ]) === "Yes"
    ) {
      await exec(`git stash push -m "OPOB Stash"`, {
        cwd: projectPath,
      });
      changes = false;
    }
  }

  if (!changes) {
    await exec(`git checkout ${branch}"`, {
      cwd: projectPath,
    });
  }

  await exec(`${editor} ${windowFlag} ${projectPath}`);
}

const cyrb53 = (str, seed = 0) => {
  let h1 = 0xdeadbeef ^ seed, h2 = 0x41c6ce57 ^ seed;
  for (let i = 0, ch; i < str.length; i++) {
    ch = str.charCodeAt(i);
    h1 = Math.imul(h1 ^ ch, 2654435761);
    h2 = Math.imul(h2 ^ ch, 1597334677);
  }
  h1 = Math.imul(h1 ^ (h1 >>> 16), 2246822507);
  h1 ^= Math.imul(h2 ^ (h2 >>> 13), 3266489909);
  h2 = Math.imul(h2 ^ (h2 >>> 16), 2246822507);
  h2 ^= Math.imul(h1 ^ (h1 >>> 13), 3266489909);

  return 4294967296 * (2097151 & h2) + (h1 >>> 0);
};

const OPOB_CACHE_FILE = kenvPath('cache/opob.json');

async function readCache() {
  return JSON.parse(await ensureReadFile(OPOB_CACHE_FILE, "{}"));
}

async function writeCache(cache) {
  await writeFile(OPOB_CACHE_FILE, JSON.stringify(cache));
}

async function getCacheKey(key) {
  const cache = await readCache();
  return cache[key] ?? null;
}

async function setCacheKey(key, value) {
  let cache = await readCache();
  cache[key] = value;
  await writeCache(cache);
}

async function getCodeEditor() {
  let editor = await getCacheKey('editor');
  if (!editor) {
    let editors = [];

    const which = await exec(`which code-insiders || which code`);
    if (which.stdout.includes("/code-insiders")) {
      editors.push("code-insiders");
    }

    if (which.stdout.includes("/code")) {
      editors.push("code");
    }

    if (editors.length > 1) {
      editor = await arg("Which Editor?", editors);
      if (await arg("Remember Editor?", ["Yes", "No"]) === 'Yes') {
        // cache the code they wish to use.
        setCacheKey('editor', editor);
      }
    }

    if(editors.length === 1) {
      editor = editors[0];
    }

    if (editors.length === 0) {
      let message = md(`# And Editor was not found
    Please install an editor CLI such as code, code-insiders or phpstorm
      `)
      await div(message);
      exit(1);
    }
  }

  return editor;
}

async function getProjects() {
    let projects = await getCacheKey('projects');
    if (!projects) {
      const p = await exec(
        `find -L ${projectDir} -maxdepth 4 -name .git -type d -prune`,
      );
      const pList = p.stdout.split("\n");
    
      projects = pList.map((project) => {
        return project.replace(projectDir + "/", "").replace("/.git", "");
      });

      await setCacheKey('projects', projects);
    }

    return projects;  
}

async function getAllProjectBranches(project) {
  let branches = getCacheKey('branches') || {};
  let projectHash = cyrb53(project);

  if (!branches[projectHash]) {
    branches[projectHash] = await fetchBranchesFromServer(project);
    await setCacheKey('branches', branches);
  }

  return currentBranchFirst(branches[projectHash]);
}

async function fetchBranchesFromServer(project) {
  let projectDir = await getProjectDirectory();
  console.log(`${projectDir}/${project}`);
  
  await exec(`cd ${projectDir}/${project}`, {
    cwd: home()
  });

  const branchString = await exec(`git fetch && git branch -a`, {
    cwd: `${projectDir}/${project}`,
  });

  let branches = branchString.stdout.split("\n");
  branches = branches.filter((b) => {
    return b !== "" && !b.includes("dependabot") &&
      !b.includes("HEAD ->") && !b.includes("* ") &&
      !branches.includes(b.replace("remotes/origin/", "").trim())
  }).map((b) => {
    return b.replace("remotes/origin/", "").trim();
  });

  return [...branches];
}

async function getCurrentBranch() {
  const current = await exec(`git symbolic-ref --short HEAD`);
  return current.stdout.trim();
}

async function currentBranchFirst(branches) {
  const currentBranch = await getCurrentBranch();

  const filteredBranches = branches.filter(b => b !== currentBranch);

  return [currentBranch, ...filteredBranches];
}

async function getProjectDirectory() {
  let projectDir = await getCacheKey('projectDir');

  if (!projectDir) {
    projectDir = await path(home());
  }

  setCacheKey('projectDir', projectDir);

  return projectDir;
}

```

**Comment by srhyne** [comment]
Bravo @johnlindquist! The new format for actions is fantastic! I struggled a bit before to get the -> right arrow to enable showing of flags. Now it works flawlessly. 

The design refresh is really great too!
